api/v1beta2/network_types.go

```diff
@@ -323,6 +323,11 @@ type VPCSpec struct {
 	// +kubebuilder:default=Ordered
 	// +kubebuilder:validation:Enum=Ordered;Random
 	AvailabilityZoneSelection *AZSelectionScheme `json:"availabilityZoneSelection,omitempty"`
+
+	// AvailabilityZoneList defines a list of Availability Zones in which to create network resources in.
+	// If defined, both AvailabilityZoneUsageLimit and AvailabilityZoneSelection are ignored.
+	// +optional
+	AvailabilityZoneList []string `json:"availabilityZoneList,omitempty"`
```

Comment:
```suggestion
	// AvailabilityZoneList defines a list of Availability Zones in which to create network resources in.
	// Cannot be defined alongside AvailabilityZoneUsageLimit and AvailabilityZoneSelection.
	// +optional
	AvailabilityZones []string `json:"availabilityZones,omitempty"`
```

We should probably refuse usage when both fields are set, that can cause confusion. We could check in a validation webhook, and given that this field is new, the change is backward compatible.

Reply:
Sounds good. 👍

Reply:
Actually, both fields have mandatory defaults. They are always defined. Meaning we cannot differentiate between deliberately set or the default value defined. Which means we cannot have a validation against them.

Reply:
We could move the defaults in the defaulting webhook instead of openapi schema? That way, the defaulting logic can be conditional.

Reply:
I'm not a 100% sure how that works. It wouldn't break existing clusters, right? Like, the default in the kubebuilder results in a different yaml, while the defaulting webhook... doesn't? Or does it?

Reply:
The defaulting webhook is translated into a mutating admission webhook by kubebuilder. New resources + resource updates will be seen by the defaulting webhook, mutated, and stored in the cluster after the mutation, but _existing_ resources will not be re-examined or updated. 

This can be a problem when the new field being introduced needs to be set to a non-nil default value for existing resources. However, that is not the case here...

So I think because the default behavior of the field is `nil` defaulting would work as intended, because existing resources are in the correct state, because none of them want to use this new functionality.

Reply:
I'm not talking about defaulting for my new API I'm talking about defaulting for existing ones. I think that's a change of behaviour I would advice against.

Reply:
We already have defaulting and validation webhooks in place, are you referring to something else?

Reply:
I'm talking about what you suggested. Where we switch the two others from using kubebuilder defaults to the validating webhook / defaulting webhook. That would be a change in behaviour I think?

Reply:
No it's not a change in behavior; the webhooks are considered part of the codebase, and so is the same the openapi/kubebuilder tags. This would move the validation from one place to the other; it's also brand new validation

Reply:
Also... Okay. So... Kubebuilder Markers GENERATE default values so the generated output will be different. But mutating admission webhooks change values in flight so the output OBJECT's value will be different.

Unless I'm super mistaken these two scenarios ARE different.

Reply:
The yaml is going to be different, although the defaulting logic moves to the webhook, which historically we have not treated as a breaking change, but as a compatible one.

Reply:
That said, what I was suggesting is to validate this new field in the webhook, and reject it when the other two colliding fields are set.

Reply:
Yes, but validating it would require removing the defaults from here. And that's what started this conversation. :)

Reply:
Yeah that's not a breaking change, @fabriziopandini @sbueringer from the upstream project can keep me honest though.